generic/MAPL_Generic.F90

```diff
@@ -1830,6 +1839,100 @@ recursive subroutine MAPL_GenericWrapper ( GC, IMPORT, EXPORT, CLOCK, RC)
 
    end subroutine MAPL_GenericWrapper
 
+   !=============================================================================
+
+   !BOPI
+
+   ! !IROUTINE: omp_driver
+
+   ! !INTERFACE:
+   subroutine omp_driver(GC, import, export, clock, RC)
+      !$ use omp_lib
+      type (ESMF_GridComp), intent(inout) :: GC     ! Gridded component
+      type (ESMF_State),    intent(inout) :: import ! Import state
+      type (ESMF_State),    intent(inout) :: export ! Export state
+      type (ESMF_Clock),    intent(inout) :: clock  ! The clock
+      integer, optional,    intent(  out) :: RC     ! Error code:
+
+      type (MAPL_MetaComp), pointer :: MAPL
+      !type(GOCART_State), pointer :: self
+      !type(wrap_) :: wrap
+      integer :: thread
+      type(ESMF_State) :: subimport
+      type(ESMF_State) :: subexport
+      integer :: status
+      integer, allocatable :: statuses(:), user_statuses(:)
+      integer :: num_threads
+      character(len=ESMF_MAXSTR) :: Iam = "Run1"
+      type(ESMF_VM) :: vm
+      type(ESMF_GridComp) :: thread_gc
+      integer :: i, me
+      integer :: userRC
+      character(len=ESMF_MAXSTR) :: comp_name
+      integer :: phase
+
+      !call start_global_time_profiler('run1()')
+      !call ESMF_UserCompGetInternalState (GC, 'GOCART_State', wrap, status)
+      !VERIFY_(status)
+      !self => wrap%ptr
+
+      call ESMF_GridCompGet (GC, vm=vm, NAME=comp_name, currentPhase=phase, __RC__)
+      call ESMF_VMGet(vm, localPet=me, __RC__)
+
+      call MAPL_GetObjectFromGC (GC, MAPL, __RC__)
+      if(MAPL%is_threading_active()) then
+         call ESMF_GridCompRun (GC, &
+              importState=import, &
+              exportState=export, &
+              clock=CLOCK, PHASE=phase, &
+              userRC=userRC, __RC__ )
+         _VERIFY(userRC)
+      else
+         !call start_global_time_profiler('activate_threads')
+         num_threads = 1
+         !$ num_threads = omp_get_max_threads()
+         call MAPL%activate_threading(num_threads, __RC__)
+         !call stop_global_time_profiler('activate_threads')
+         !call start_global_time_profiler('parallel')
+
+         allocate(statuses(num_threads), __STAT__)
+         allocate(user_statuses(num_threads), __STAT__)
+         statuses=0
+         user_statuses=0
+         !$omp parallel default(none), &
+         !$omp& private(thread, subimport, subexport, thread_gc), &
+         !$omp& shared(gc, statuses, user_statuses, clock, PHASE, MAPL)
+
+         thread = 0
+         !$ thread = omp_get_thread_num()
```

Comment:
Do we maybe want to bury this construct into something like `MAPL_get_thread_id()` or something like that (not sure of best name).   It's otherwise a bit jarring to realize the "commented" line overrules the line before it.

Reply:
We might right a `MAPL_get_thread_id()` but won't the same two lines be present in that function too?

Reply:
Yes, but ... then at least it is only in _one_ place instead of scattered in a few places.   Plus one can document the odd construct in the helper procedure, whereas documenting each 2 line bit in the code would further undermine legibility.

Reply:
Got it. Should this be part of MetaComp?

Reply:
Ideally, we want to push as much logic into the `OpenMP_Support` layer as possible.   Meta is already too large.

Reply:
Okay.

Reply:
Since it will be in `OpenMP_Support` instead of MAPL per se, I am about to name it `get_current_thread_id`. Is this okay?

Reply:
I will add another function to get num_threads for the same reason.